### ISSUE TITLE
OAI_ID consistency required for IRUS

### ISSUE THREAD

**Issue by grahamjevon** [scientist-softserv/britishlibrary#380]
@cziaarm As I mentioned last week, I encountered an error when reporting on item requests in IRUS (Custom Platform Master Report and Combined R4/R5 Platform Report). I am able to generate reporting data for investigations, but not for requests. I notified JISC. The response from JISC below (Ref:IN:01357909) identifies the problem and solution:

Hi Graham

Apologies for the delay in replying, Easter got in the way!

We've had a poke around our system and found that the problem is that IRO repositories are exposing different OAI_IDs for an item depending on whether the usage for it is a metadata view (Investigation) or a full-content/file download (Request).

During IRUS processing we attempt to harvest metadata for each item, and the resulting statistics only include usage for the items for which those requests are successful.  We're seeing metadata for Investigations but not Requests.

Looking at one example, "The Spiral-Locked Letters of Elizabeth I and Mary, Queen of Scots", an Investigation is registered for a visit to https://bl.iro.bl.uk/concern/articles/a78e939c-4fc1-4217-bae2-6ac18623b7e4, as metadata is successfully returned from https://bl.iro.bl.uk/catalog/oai?verb=GetRecord&metadataPrefix=oai_dc&identifier=oai:bl:a78e939c-4fc1-4217-bae2-6ac18623b7e4.  However, the full-text of the item is accessed via https://bl.iro.bl.uk/downloads/77335a38-9758-4d87-902a-b3d0298282de, and a metadata request for this, https://bl.iro.bl.uk/catalog/oai?verb=GetRecord&metadataPrefix=oai_dc&identifier=oai:bl:77335a38-9758-4d87-902a-b3d0298282de, gives an 'idDoesNotExist' error.

The solution is for IRO to expose the same OAI_ID for an item, regardless of whether it’s a metadata view or download of the associated full content, rather than showing different identifiers.

Hope that helps.

Cheers
Dave

**Comment by cziaarm** [comment]
Thanks Graham,

I honestly thought that sending different IDs would be an improvement.

Seems that other (i.e. EPrints) will send only the record ID even when they are reporting the download of a specific document or file. The record ID may represent multiple documents/files which is why I would have thought being more specific would be a bonus... but I guess not

I'll take a look at how to send the parent record ID on file download, hopefully this will be achievable without a performance impact.

Will update here soon.

Thanks,

Rory

**Comment by ShanaLMoore** [comment]
Hi @cziaarm 

We noticed you moved this one to our QA, even though it's assigned to your team. 

If this was intentional, please update this ticket with testing instructions. Jeremy reviewed this on a code level but did not manually test it. 

Otherwise, would it be ok to deploy to staging and have graham re-review?

**Comment by cziaarm** [comment]
Yep, I moved it to QA so we could QA it before moving it on for Graham and the BL. So no action needed from you guys on this.

Thanks,

Rory

**Comment by cziaarm** [comment]
Tested with https://bl.bl-staging.notch8.cloud/concern/articles/95f6017f-3539-46f0-b09b-095dedd474da?locale=en

Work show page:

INFO -- : [ActiveJob] [IrusAnalytics::IrusClient] [4235ab89-fd78-4799-92af-729c8e6eceeb] Performing IrusAnalytics::IrusClient (Job ID: 4235ab89-fd78-4799-92af-729c8e6eceeb) from Sidekiq(irus_analytics) with arguments: "https://irus.jisc.ac.uk/counter/", {:date_stamp=>"2023-06-01T15:25:46Z", :client_ip_address=>"81.174.241.126", :user_agent=>"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36", **:item_oai_identifier=>"oai:hyku:95f6017f-3539-46f0-b09b-095dedd474da"**, :file_url=>"https://bl.bl-staging.notch8.cloud/concern/articles/95f6017f-3539-46f0-b09b-095dedd474da?locale=en", :http_referer=>"https://bl.bl-staging.notch8.cloud/", :source_repository=>"iro.bl.uk"}, "Investigation"


Download of file: 

INFO -- : [ActiveJob] [IrusAnalytics::IrusClient] [f976936f-175e-4a44-a455-a04ff245fffb] Performing IrusAnalytics::IrusClient (Job ID: f976936f-175e-4a44-a455-a04ff245fffb) from Sidekiq(irus_analytics) with arguments: "https://irus.jisc.ac.uk/counter/", {:date_stamp=>"2023-06-01T15:26:07Z", :client_ip_address=>"81.174.241.126", :user_agent=>"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36", **:item_oai_identifier=>"oai:hyku:95f6017f-3539-46f0-b09b-095dedd474da"**, :file_url=>"https://bl.bl-staging.notch8.cloud/downloads/c818a42c-3937-4ef1-b490-d8917ebddf3b?locale=en", :http_referer=>"https://bl.bl-staging.notch8.cloud/concern/articles/95f6017f-3539-46f0-b09b-095dedd474da?locale=en", :source_repository=>"iro.bl.uk"}, "Request"


FileSet show page (actually only available for logged in users of course):

INFO -- : [ActiveJob] [IrusAnalytics::IrusClient] [d29a3f89-b782-42ff-9616-d03a605f1547] Performing IrusAnalytics::IrusClient (Job ID: d29a3f89-b782-42ff-9616-d03a605f1547) from Sidekiq(irus_analytics) with arguments: "https://irus.jisc.ac.uk/counter/", {:date_stamp=>"2023-06-01T15:29:18Z", :client_ip_address=>"81.174.241.126", :user_agent=>"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36", **:item_oai_identifier=>"oai:hyku:95f6017f-3539-46f0-b09b-095dedd474da"**, :file_url=>"https://bl.bl-staging.notch8.cloud/concern/file_sets/c818a42c-3937-4ef1-b490-d8917ebddf3b?locale=en", :http_referer=>"https://bl.bl-staging.notch8.cloud/concern/articles/95f6017f-3539-46f0-b09b-095dedd474da?locale=en", :source_repository=>"iro.bl.uk"}, "Investigation"
